l10n_it_delivery_note/models/stock_delivery_note.py

```diff
@@ -546,6 +549,61 @@ def action_confirm(self):
                 ).next_by_id()
                 note.sequence_id = sequence
 
+    def action_confirm(self):
+        for note in self:
+            warning_message = False
+            carrier_ids = note.mapped("picking_ids.carrier_id")
+            transporter_ids = carrier_ids.mapped("partner_id")
+            if len(transporter_ids) > 1:
+                warning_message = _(
+                    "This delivery note contains pickings related to different transporters. "
+                    "Are you sure you want to proceed? \nTransporters: %(transporters)s",
+                    transporters=", ".join(transporter_ids.mapped("name")),
+                )
+            elif (
+                transporter_ids
+                and note.carrier_id
+                and note.carrier_id != transporter_ids
+            ):
+                warning_message = _(
+                    "The carrier set in Delivery Note is different "
+                    "from the carrier set in picking(s). "
+                    "Are you sure you want to proceed?"
+                )
+            elif len(carrier_ids) > 1:
+                warning_message = _(
+                    "This delivery note contains pickings related to different "
+                    "delivery methods from the same transporter. "
+                    "Are you sure you want to proceed? \nDelivery Methods: %(carriers)s",
+                    carriers=", ".join(carrier_ids.mapped("name")),
+                )
+            elif (
+                carrier_ids
+                and note.delivery_method_id
+                and carrier_ids != note.delivery_method_id
+            ):
+                warning_message = _(
+                    "The shipping method set in Delivery Note is different "
+                    "from the shipping method set in picking(s). "
+                    "Are you sure you want to proceed?"
+                )
+            if warning_message:
+                return {
+                    "type": "ir.actions.act_window",
+                    "name": _("Warning"),
+                    "res_model": "stock.delivery.note.confirm.wizard",
+                    "view_type": "form",
+                    "target": "new",
+                    "view_mode": "form",
+                    "context": {
+                        "default_delivery_note_id": note.id,
+                        "default_warning_message": warning_message,
+                        **self._context,
+                    },
+                }
+            else:
+                note._action_confirm()
```

Comment:
Question: È possibile che alcuni di questi warning vadano mostrati insieme? 

Ad esempio: Il DN fa riferimento a diversi carrier, ma anche a diversi metodi di spedizione. In tal caso, pensi che avrebbe senso aggiungere al messaggio di warning tutti i warning necessari? (con un piccolo refactor sulle condizioni)

Per come è adesso, viene mostrato solo il primo, e se confermi gli altri non vengono processati.

Reply:
@aleuffre a livello logico ha senso credo, perchè: 

- se il DN è relazionato a picking con carrier diversi, ne consegue che il carrier nel DN sarà diverso da almeno uno dei carrier nel picking
- se il DN è relazionato a picking con metodi diversi, ne consegue che il metodo nel DN sarà diverso da almeno uno dei metodi nel picking
- se invece tutti i picking hanno lo stesso carrier ma è diverso quello del DN, viene mostrato il warning relativo a questa discrepanza
- stesso comportamento del punto precedente per quanto riguarda i metodi di consegna

Mi confermi che il comportamento della PR è quello che ho descritto?

Reply:
Si, confermo, a parte il fatto che il secondo e il terzo controllo sono in ordine inverso rispetto a come li hai scritti tu.

In ogni caso non credo sia un problema, mi hai convinto

Reply:
Sistemato l'ordine dei warning, grazie per la verifica